### PR TITLE
[REVIEW] Harden inconsistent-schema handling in pyarrow-based read_parquet

### PR DESCRIPTION
**Pull request by rjzamora**
Addressed [cudf#5062](https://github.com/rapidsai/cudf/issues/5062)

The `ArrowEngine` version of `read_parquet` is meant to handle inconsistent schemas if the user passes in `dataset={"validate_schema": False}`.  However, this is currently crashing in some cases when the dataset is written with `dask_cudf`.  This PR adds a small fix to make sure the schema is defined when the `pq.ParquetDataset` API fails to populate it.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by rjzamora** [comment]
@martindurant @TomAugspurger - Any thoughts on this change?

File: dask/dataframe/io/tests/test_parquet.py [hunk 1]
```diff
@@ -2044,6 +2044,40 @@ def test_timeseries_nulls_in_schema_pyarrow(tmpdir, timestamp, numerical):
...
+    df1 = pd.DataFrame({"id": [0, 1], "val": [10, 20]})
+    df2 = pd.DataFrame({"id": [2, 3], "val": [30, 40]})
+
+    desired_type = "int64"
+    other_type = "int32"
+    df1.val = df1.val.astype(desired_type)
+    df2.val = df2.val.astype(other_type)
+
+    df_expect = pd.concat([df1, df2], ignore_index=True)
+    df_expect.val = df_expect.val.astype(desired_type)
```
Comment (TomAugspurger) [inline dask/dataframe/io/tests/test_parquet.py:2065]: Just to be safe

```suggestion
    df_expect['val'] = df_expect.val.astype(desired_type)
```

Is this line necessary for pandas? I'd expect it to upcast to int64 already (Happy to keep it consistent if it was needed for cudf, mostly just curious).

File: dask/dataframe/io/tests/test_parquet.py [hunk 2]
```diff
@@ -2044,6 +2044,40 @@ def test_timeseries_nulls_in_schema_pyarrow(tmpdir, timestamp, numerical):
...
+    df1 = pd.DataFrame({"id": [0, 1], "val": [10, 20]})
+    df2 = pd.DataFrame({"id": [2, 3], "val": [30, 40]})
+
+    desired_type = "int64"
+    other_type = "int32"
+    df1.val = df1.val.astype(desired_type)
+    df2.val = df2.val.astype(other_type)
+
+    df_expect = pd.concat([df1, df2], ignore_index=True)
+    df_expect['val'] = df_expect.val.astype(desired_type)
```
Comment (TomAugspurger) [inline dask/dataframe/io/tests/test_parquet.py:2065]: Sorry, my suggestioned introduced a lint error :) Let's try this.

```suggestion
    df_expect["val"] = df_expect.val.astype(desired_type)
```

**Comment by martindurant** [comment]
Thanks for this

**Pull request merged**